File: Practice04-Dgemm.py
import numpy as np
import time
import math

l_max = 50
# creation of empty array ttime
ttime = np.empty(l_max, dtype="int")

# reading alpha,beta,op(A),op(B),C from 'input.txt'
f = open('input.txt')


alpha = int(f.readline())
beta = int(f.readline())
m = int(f.readline())
k = int(f.readline())
n = int(f.readline())

Array_op_A = np.empty((m, k), dtype="int")
Array_op_B = np.empty((k, n), dtype="int")
Array_C = np.empty((m, n), dtype="int")
Array_D = np.empty((m, n), dtype="int")

i = 0
while i < m:
     j = 0
     while j < k:
         Array_op_A[i][j] = int(f.readline())
#         print(str(Array_op_A[i][j]))
         j = j + 1
     i = i + 1


i = 0
while i < k:
     j = 0
     while j < n:
         Array_op_B[i][j] = int(f.readline())
#        print("B"+str(i)+str(j)+"   "+str(Array_op_B[i][j]))
         j = j + 1
     i = i + 1

i = 0
while i < m:
    j = 0
    while j < n:
        Array_C[i][j] = int(f.readline())
#       print("C" + str(i) + str(j) + "   " + str(Array_C[i][j]))
        j = j + 1
    i = i + 1

f.close()


l = 0
while l < l_max:

#    t0 = time.perf_counter_ns()
#    t0 = time.time()
    t0 = time.time_ns()

# calculation matrix D

    i = 0
    while i < m:
        j = 0
        while j < n:
             Array_D[i][j] = 0
             i1 = 0
             while i1 < k:
                Array_D[i][j] = Array_D[i][j] + Array_op_A[i][i1] * Array_op_B[i1][j]
                i1 = i1 + 1
#            print("D" + str(i) + str(j) + "   " + str(Array_D[i][j]))
             Array_D[i][j] = alpha * Array_D[i][j] + beta * Array_C[i][j]
             j = j + 1
        i = i + 1

        #   t1 = time.time()
    t1 = time.time_ns()
        #  t1 = time.perf_counter_ns()
    ttime[l] = (t1 - t0)/1000

    print("l=" + str(l) + "  time=" + str(ttime[l]))

#    i = 0
#    while i < m:
#        j = 0
#        while j < n:
#            print(" " + "(" + str(i) + str(j) + ") " + str(Array_D[i][j]) + " ")
#           j = j + 1
#        i = i + 1
#        print()

    l = l + 1

# sorting array ttime
ttime.sort()
print()

# writing sorted array ttime in the file 'time.txt'
file_time = open('time.txt', 'w')
file_time.write(str(l_max) + "\n")
ll = 0
while ll < l_max:
    file_time.write(str(ll) + "\n")
    file_time.write(str(ttime[ll]) + "\n")
    ll = ll + 1

# calculation average value
av = 0.
ll = 0
while ll < l_max:
    av = av + float(ttime[ll])
    print("l=" + str(ll) + "    " + "t=" + str(ttime[ll]) +  "    " + str(av))
    ll = ll + 1

av = av / float(l_max)

# calculation mean deviation

s = 0
l1 = 0
while l1 < l_max:
    s = s + math.pow(av - ttime[l1], 2)
    l1 = l1 + 1

Dev = math.sqrt(s / l_max)

# calculation median value

if int(l_max) % 2 != 0:
    l1 = int((l_max - 1) / 2)
    print("l1=" + str(l1))
    med = ttime[l1]

else:
    l2 = int(l_max / 2-1)
    l3 = int(l_max / 2)
    print("l2=" + str(l2) + "    l3=" + str(l3))
    med = (ttime[l2] + ttime[l3]) / 2.


print("shortest time= " + str(ttime[0]) + "\n" + "longest time= " + str(ttime[l_max - 1]))
print("average value= " + str(av))
print("median value= " + str(med))
print("mean deviation= " + str(Dev))

File: Practice04-Dgemm.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DGEMM_int
{
    internal class Program
    {
        static void Main(string[] args)
        {
            int l_max = 50;
            int alpha;
            int beta;

            long[] time = new long[l_max];

            int m, k, n;

            int i, j;
            long t0, t1;


            try
            {

                StreamReader sr = File.OpenText("input.txt");

// reading alpha, beta, op(A), op(B), C from Input.txt

                alpha = Int32.Parse(sr.ReadLine());
                beta = Int32.Parse(sr.ReadLine());
                m = Int32.Parse(sr.ReadLine());
                k = Int32.Parse(sr.ReadLine());
                n = Int32.Parse(sr.ReadLine());

              

                int[,] Array_op_A = new int[m, k];
                int[,] Array_op_B = new int[k, n];
                int[,] Array_C = new int[m, n];
                int[,] Array_D = new int[m, n];

                for (i = 0; i < m; i++)
                    for (j = 0; j < k; j++)
                    {
                        Array_op_A[i, j] = Int32.Parse(sr.ReadLine());
                      //  Console.WriteLine(Array_op_A[i, j]);
                    }

                for (i = 0; i < k; i++)
                    for (j = 0; j < n; j++)
                    {
                        Array_op_B[i, j] = Int32.Parse(sr.ReadLine());
                      //  Console.WriteLine(Array_op_B[i, j]);
                    }

                for (i = 0; i < m; i++)
                    for (j = 0; j < n; j++)
                    {
                        Array_C[i, j] = Int32.Parse(sr.ReadLine());
                       // Console.WriteLine(Array_C[i, j]);
                    }

                sr.Close();

                for (int l = 0; l < l_max; l++)
                {                

                    DateTime date_time0 = DateTime.Now;
                    t0 = date_time0.Millisecond;

                   // t0 = GetNanoseconds();
                   // Console.WriteLine("t0="+t0);
                  

                    for (i = 0; i < m; i++)
                        for (j = 0; j < n; j++)
                        {
                            Array_D[i, j] = 0;
                            for (int i1 = 0; i1 < k; i1++)
                                Array_D[i, j] = Array_D[i, j] + Array_op_A[i, i1] * Array_op_B[i1, j];
                            Array_D[i, j] = alpha * Array_D[i, j] + beta * Array_C[i, j];
                            //   Array_C[i][j] = Array_D[i][j];
                        }
                 
                                  
                    DateTime date_time1 = DateTime.Now;
                    t1  = date_time1.Millisecond;

                    // t1 = GetNanoseconds();
                    //  Console.WriteLine("t1=" + t1);

                    time[l] = t1 - t0;
                    /**
                     // print matrix D
                     System.out.print("l=" + l + "  time=" + time[l] + "\n");
                    for (i = 0; i < m; i++) {  
                        for (j = 0; j < n; j++) {
                            System.out.print(" " + "(" + i + j + ") " + Array_D[i][j] + " "); 
                        }
                        System.out.println();
                    }
                     */
                }

            }
            catch (IOException ex)
            {
            }


         // sort array 'time'
            Array.Sort(time);


// write array 'time' to file 'time.txt'
            StreamWriter swTime = new StreamWriter("time.txt");

            String string1 = l_max.ToString();
            swTime.WriteLine(string1);
            for (int l = 0; l < l_max; l++)
            {
                String string2 = l.ToString();
                String string3 = time[l].ToString();
                swTime.WriteLine(string2 + "\n" + string3);
            }
            swTime.Close();


            Console.WriteLine();

            // calculation the average value
            double av = 0;
            for (int l = 0; l < l_max; l++)
            {
                av = av + (double)time[l];
                Console.WriteLine("l=" + l + "    " + "t=" + time[l] + " nanosec");
            }
            av = av / l_max;

// calculation the mean dispersion
            double s = 0;
            for (int l = 0; l < l_max; l++)
            {
                s = s + Math.Pow(av - (double)time[l], 2);
            }

            double Dev = Math.Sqrt(s / l_max);

// calculation the median value
            double med;
            if (l_max % 2 != 0)
            {
                int l1 = (l_max - 1) / 2;
                med = time[l1];
                //  System.out.println("l1=" + l1);
            }
            else
            {
                int l2 = l_max / 2 - 1;
                int l3 = l_max / 2;
                med = (double)(time[l2] + time[l3]) / 2d;
                //  System.out.println("l2=" + l2+"    l3=" + l3 );
            }

            Console.WriteLine("shortest time= " + time[0] + "\n" + "longest time= " + time[l_max - 1]);
            Console.WriteLine("average value= " + av);
            Console.WriteLine("median value= " + med);
            Console.WriteLine("mean deviation= " + Dev);
            Console.ReadLine();

        }

        public static long GetNanoseconds()
        {
            double timestamp = Stopwatch.GetTimestamp();
            double nanoseconds = 1_000_000_000.0 * timestamp / Stopwatch.Frequency;

            return (long)nanoseconds;
        }


    }
}






